### PR TITLE
Global Message Passing Layer for MXMNet Model

### PR DESCRIPTION
**Pull request by riya-singh28**
## Description

Implemented a global Message Passing Layer for MXMNet Model.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by riya-singh28** [comment]
image

**Comment by riya-singh28** [comment]
@rbharath @vsomnath @ARY2260 PR is ready to review.

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -4187,3 +4189,200 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+    """
+
+    def __init__(self, config: dict):
+        """Initializes the GlobalMessagePassing layer.
+
+        Parameters
+        -----------
+        config: dict
+            A dictionary containing the configuration parameters.
+        dim: int
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:4262]: only mention parameters taken input during __init__ call.

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -4187,3 +4189,200 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+    torch.Size([5, 1])
+
+    """
+
+    def __init__(self, config: dict):
+        """Initializes the GlobalMessagePassing layer.
+
+        Parameters
+        -----------
+        config: dict
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:4260]: Instead of config, it's better to use the keys in the ` config` as parameters. It might get confusing to use this layer later on.

**Review by ARY2260 — CHANGES_REQUESTED** [review]

File: deepchem/models/tests/test_layers.py [hunk 3]
```diff
@@ -1183,3 +1184,43 @@ def test_edge_network():
...
+@pytest.mark.torch
+def test_global_MP():
+
+    seed = 123
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+    np.random.seed(seed)
+    random.seed(seed)
```
[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/tests/test_layers.py:1198]: Yes, the test passes without setting the random seed. Removed random seed.

File: deepchem/models/tests/test_layers.py [hunk 4]
```diff
@@ -1183,3 +1184,43 @@ def test_edge_network():
...
+                              [-0.2644], [-0.2644], [-0.2644]])
+
+    edge_indices = torch.tensor(
+        [[0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4],
+         [1, 2, 3, 4, 0, 2, 3, 4, 0, 1, 3, 4, 0, 1, 2, 4, 0, 1, 2, 3]])
+
+    out = dc.models.torch_models.layers.GlobalMessagePassing(config)
+    output = out(node_features, edge_attr, edge_indices)
+    output = output.detach().numpy()
+    assert np.allclose(
```
[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/tests/test_layers.py:1222]: okay, I'll assert it directly.

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -4187,3 +4189,200 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
         output = output / torch.sum(output, dim=2, keepdim=True)
         output = output.view(-1, self.n_input * 11)
         return output
+
+
+class ResidualBlock(nn.Module):
```
[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4194]: @rbharath Do you want me to add ResidualBlock as a separate layer or a helper class just for MXMNet layers?

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -4187,3 +4189,200 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+        torch.Tensor
+            The output tensor after applying the residual block.
+        """
+        x_residual: torch.Tensor = self.mlp(x)
+        x_out: torch.Tensor = x_residual + x
+        return x_out
+
+
+class GlobalMessagePassing(MessagePassing):
+    """This class implements the Global Message Passing Layer from the Molecular Mechanics-Driven Graph Neural Network
```
[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4232]: Done

File: deepchem/models/torch_models/layers.py [hunk 7]
```diff
@@ -4187,3 +4189,200 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+    with Multiplex Graph for Molecular Structures(MXMNet) paper.
+
+    Examples
```
[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4235]: added

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4262]: okay

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4260]: Okay. I'll do that

File: deepchem/models/torch_models/layers.py [hunk 8]
```diff
@@ -4187,3 +4189,200 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+        --------
+        torch.Tensor
+            The output tensor after applying the residual block.
+        """
+        x_residual: torch.Tensor = self.mlp(x)
+        x_out: torch.Tensor = x_residual + x
+        return x_out
+
+
+class GlobalMessagePassing(MessagePassing):
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/torch_models/layers.py:4232]: Yes agreed, let's rename it as you suggest to prevent namespace conflict

File: deepchem/models/torch_models/layers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/torch_models/layers.py:4194]: A separate layer I think. Please go ahead and add the usage example

File: deepchem/models/tests/test_layers.py [hunk 9]
```diff
@@ -1183,3 +1183,35 @@ def test_edge_network():
         np.array(torch_result),
         np.load("deepchem/models/tests/assets/edgenetwork_result.npy"),
         atol=1e-04)
+
+
+@pytest.mark.torch
+def test_global_MP():
```
Comment (vsomnath) [inline deepchem/models/tests/test_layers.py:1189]: `test_global_message_passing`

File: docs/source/api_reference/torch_layers.csv [hunk 10]
```diff
@@ -40,3 +40,4 @@ MolGANAggregationLayer,`ref <https://arxiv.org/abs/1805.11973>`_, MolGan
 DTNNEmbedding, ref`<https://arxiv.org/abs/1609.08259>`_, DTNNModel
 DTNNStep, ref`<https://arxiv.org/abs/1609.08259>`_, DTNNModel
 DTNNGather, ref`<https://arxiv.org/abs/1609.08259>`_, DTNNModel
+GlobalMessagePassing, `ref <https://arxiv.org/pdf/2011.07457.pdf>`_, Global Message Passing Neural Network
```
Comment (vsomnath) [inline docs/source/api_reference/torch_layers.csv:43]: Change the reference to the `https://arxiv.org/abs/2011.07457` instead of the pdf version to maintain consistency with the rest.

File: deepchem/models/torch_models/layers.py [hunk 11]
```diff
@@ -4187,3 +4189,207 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+            The target node features tensor of shape (num_edges, feature_dim).
+        edge_attributes: torch.Tensor
+            The edge attribute tensor of shape (num_edges, attribute_dim).
+
+        Returns
+        --------
+        torch.Tensor
+            The constructed messages tensor.
+        """
+        num_edge: int = edge_attr.size()[0]
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:4386]: Is there a case where `x_i` will have a different size at dimension 0 compared to `edge_attr`? I am trying to understand the thought process behind `num_edge`

File: deepchem/models/torch_models/layers.py [hunk 12]
```diff
@@ -4187,3 +4189,207 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+        super(GlobalMessagePassing, self).__init__()
+        self.dim = dim
+
+        self.h_mlp: MultilayerPerceptron = MultilayerPerceptron(
+            d_input=self.dim, d_output=self.dim, activation_fn='silu')
+
+        self.res1: ResidualBlock = ResidualBlock(self.dim)
+        self.res2: ResidualBlock = ResidualBlock(self.dim)
+        self.res3: ResidualBlock = ResidualBlock(self.dim)
+        self.mlp: MultilayerPerceptron = MultilayerPerceptron(
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:4313]: Making the activation argument general instead of `silu` as advised by @ARY2260.

File: deepchem/models/torch_models/layers.py [hunk 13]
```diff
@@ -4187,3 +4189,207 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+            d_input=self.dim, d_output=self.dim, activation_fn='silu')
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:4314]: Same as above

File: deepchem/models/torch_models/layers.py [hunk 14]
```diff
@@ -4187,3 +4189,207 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+
+        self.x_edge_mlp: MultilayerPerceptron = MultilayerPerceptron(
+            d_input=self.dim * 3, d_output=self.dim, activation_fn='silu')
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:4317]: Same as above

**Review by vsomnath — CHANGES_REQUESTED** [review]
I have left some initial comments. Will check the equations next and leave a second pass once these smaller changes from Aryan and myself are addressed.

File: deepchem/models/torch_models/layers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4194]: I found out that by making a little change in MultilayerPerceptron, it can be used as ResidualBlock as well. I have made a PR for that, can you please review it? 
@rbharath @vsomnath @ARY2260

[the remark this answers is not in the document]
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4194]: Link - https://github.com/deepchem/deepchem/pull/3494

File: deepchem/models/tests/test_layers.py (hunk 9 above)
Comment (riya-singh28) [reply deepchem/models/tests/test_layers.py:1189]: okay, I'll do.

File: deepchem/models/torch_models/layers.py (hunk 12 above)
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4313]: Sure, I'll do.

File: deepchem/models/torch_models/layers.py (hunk 11 above)
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4389]: Actually there was a mistake, the shape of `x_i` and `x_j` is (num_edges+num_nodes, feature_dim) due to the addition of self_loops. Variable `num_edge` is equal to num_edges and `x_i[:num_edge],` `x_j[:num_edge],` and `edge_attr` is being concatenated. While the self_loops are being handled separately.

File: docs/source/api_reference/torch_layers.csv (hunk 10 above)
Comment (riya-singh28) [reply docs/source/api_reference/torch_layers.csv]: okay, I'll do

File: deepchem/models/torch_models/layers.py (hunk 11 above)
Comment (vsomnath) [reply deepchem/models/torch_models/layers.py:4389]: Ah sure, that makes sense!

File: deepchem/models/tests/test_layers.py [hunk 15]
```diff
@@ -1183,3 +1183,36 @@ def test_edge_network():
         np.array(torch_result),
         np.load("deepchem/models/tests/assets/edgenetwork_result.npy"),
         atol=1e-04)
+
+
+@pytest.mark.torch
+def test_global_message_passing():
+
```
Comment (ARY2260) [inline deepchem/models/tests/test_layers.py:1190]: add a docstring describing the test

Comment (riya-singh28) [reply deepchem/models/tests/test_layers.py]: Sure, I'll do.

File: deepchem/models/tests/test_layers.py [hunk 16]
```diff
@@ -1204,3 +1204,37 @@ def test_edge_network():
...
+    edge_indices = torch.tensor(
+        [[0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4],
+         [1, 2, 3, 4, 0, 2, 3, 4, 0, 1, 3, 4, 0, 1, 2, 4, 0, 1, 2, 3]])
+
+    out = dc.models.torch_models.layers.MXMNetGlobalMessagePassing(
+        dim, activation_fn='silu')
+    output = out(node_features, edge_attr, edge_indices)
+    output = output.detach().numpy()
+    result = np.array([[-0.27947044], [2.417905], [2.417905], [2.4178727],
```
Comment (vsomnath) [inline deepchem/models/tests/test_layers.py:1236]: How did you compute this result? Was it done by hand?

**Comment by vsomnath** [comment]
I checked the implemented equations against the original codebase and figure from the paper. The equations look good to me - have just asked one clarification about the test

File: docs/source/api_reference/layers.rst [hunk 17]
```diff
@@ -233,6 +233,10 @@ Torch Layers
 
 .. autoclass:: deepchem.models.torch_models.layers.WeaveGather
   :members:
+  
+.. autoclass:: deepchem.models.torch_models.layers.MXMNetGlobalMessagePassing
+   :members:
```
[the remark this answers is not in the document]
Comment (riya-singh28) [reply docs/source/api_reference/layers.rst]: Thank you for pointing it out. I have corrected it now.

File: deepchem/models/tests/test_layers.py (hunk 16 above)
Comment (riya-singh28) [reply deepchem/models/tests/test_layers.py:1236]: After checking that the architecture of the layer implemented and the one in the original code base are same. I just asserted the result I got.
I tried using manual_seed to exactly compare both the layers side by side. But though the implementation of MLP in original code base is through list comprehension and is different from one in deepchem, the weights are initialized differently. I wanted to discuss about it in the meeting. Can you suggest something for this?

**Comment by riya-singh28** [comment]
image

**Comment by riya-singh28** [comment]
image

**Comment by ARY2260** [comment]
> After checking that the architecture of the layer implemented and the one in the original code base are same. I just asserted the result I got. I tried using manual_seed to exactly compare both the layers side by side. But though the implementation of MLP in original code base is through list comprehension and is different from one in deepchem, the weights are initialized differently. I wanted to discuss about it in the meeting. Can you suggest something for this?

I am waiting for this discussion in office hours meeting, before finalizing review. Also, please rebase the PR.

**Review by ARY2260 — CHANGES_REQUESTED** [review]
> After checking that the architecture of the layer implemented and the one in the original code base are same. I just asserted the result I got. I tried using manual_seed to exactly compare both the layers side by side. But though the implementation of MLP in original code base is through list comprehension and is different from one in deepchem, the weights are initialized differently. I wanted to discuss about it in the meeting. Can you suggest something for this?

I am waiting for this discussion in office hours meeting, before finalizing review. Also, please rebase the PR.

**Comment by riya-singh28** [comment]
> > After checking that the architecture of the layer implemented and the one in the original code base are same. I just asserted the result I got. I tried using manual_seed to exactly compare both the layers side by side. But though the implementation of MLP in original code base is through list comprehension and is different from one in deepchem, the weights are initialized differently. I wanted to discuss about it in the meeting. Can you suggest something for this?
> 
> I am waiting for this discussion in office hours meeting, before finalizing review. Also, please rebase the PR.

There was some confusion related to seeding. And I realised that I was wrong, both the MLPs are working same and giving same output regardless of how are they implemented. I have verified the Global Message Passing layer with the author's implementation, and both are giving same output.
@rbharath @vsomnath @ARY2260 Can you please re-review it?

**Review by ARY2260 — APPROVED** [review]

**Comment by ARY2260** [comment]
LGTM
